cpp/src/arrow/flight/sql/protocol_internal.h

```diff
@@ -24,3 +24,4 @@
 #include "arrow/flight/sql/visibility.h"
 
 #include "arrow/flight/sql/FlightSql.pb.h"  // IWYU pragma: export
+#include "arrow/flight/Flight.pb.h"
```

Comment:
@lidavidm re: https://github.com/apache/arrow/pull/34817/files#r1245582470 not exactly sure of preference here, whether we consider the Flight messages to be part of the FSQL protocol, or include both protocols in the FSQL client/server modules (more noisy but...).  Preference?

Reply:
Sorry, I missed this; I think you won't be able to do this because of Protobuf linking complications. Instead, you'll have to define a C++ wrapper type that gets exported from libarrow_flight.dll, which gets used in libarrow_flight_sql.dll (i.e. Flight SQL won't be able to know/see that they're actually Protobufs underneath)

Reply:
Not super familiar with how the Windows linkage works here, do you mean:

* Façade(?) (presumable with an MVP set of functionality passed through from the Proto types) in libarrow_flight, i.e. in... flight/protocol_internal.h or another (new?) module?
* Including the base Flight protocol_internal.h in sql/protocol_internal.h (to expose the symbols for dynamic import)?
* And I assume no changes required to add a dynamic dependency of libarrow_flight on libarrow_flight_sql (which already exists?)

TIA!

Reply:
Basically: you cannot expose any Protobuf types in an Arrow API, implicitly or explicitly. All Protobuf messages need to have an equivalent, pure-C++ definition.

Reply:
So I guess (despite our discussion about keeping all the handler code in fSQL) that means having the Proto ser/de in libarrow itself?  Trying to think how this is going to work given where ParseAction*Request() and PackActionResult() calls are—I guess even without handlers being there the Parse/Pack functions can live in libarrow, including the overloaded PackActionResult, although that in turn doesn't have visibility of the generic PackActionResult (if it's in libarrow, which otherwise has no concept of Action I believe) so... any thoughts on how that's actually going to end up laid out to solve the visibility & linkage issues in a sane/tidy way?

Reply:
As in another thread this has been updated to use new code in serialization_internal